File: archive/internal/protocol/codec/api/init_producer_id.go
package api

import "github.com/woestebanaan/skafka/internal/protocol/codec"

// InitProducerIdRequest (key 22, v0–v4). The Java client sends this on
// startup when enable.idempotence=true (the default since Kafka 3.0)
// to obtain a (producerId, epoch) pair used to tag every Produce batch
// for the lifetime of the producer.
//
// v3+ adds (ProducerID, ProducerEpoch) request fields per KIP-360 so a
// producer can ask the broker to renew its existing PID after a fatal
// error (instead of allocating a fresh one and risking duplicate
// records). skafka currently treats every InitProducerId as fresh —
// the request fields are accepted but ignored.
//
// Flexibility: v0/v1 use the legacy header (no tagged fields); v2+
// uses REQUEST_HEADER_V2 with tagged fields.
type InitProducerIdRequest struct {
	TransactionalID      string // v0+, nullable; "" + null=true ⇒ non-transactional producer
	TransactionTimeoutMs int32  // v0+
	ProducerID           int64  // v3+
	ProducerEpoch        int16  // v3+
}

// InitProducerIdResponse (key 22, v0–v4).
type InitProducerIdResponse struct {
	ThrottleTimeMs int32
	ErrorCode      int16
	ProducerID     int64
	ProducerEpoch  int16
}

func DecodeInitProducerIdRequest(r *codec.Reader, version int16) (*InitProducerIdRequest, error) {
	req := &InitProducerIdRequest{ProducerID: -1, ProducerEpoch: -1}
	flexible := version >= 2

	var err error
	req.TransactionalID, _, err = nullableString(r, flexible)
	if err != nil {
		return nil, err
	}
	if req.TransactionTimeoutMs, err = r.ReadInt32(); err != nil {
		return nil, err
	}
	if version >= 3 {
		if req.ProducerID, err = r.ReadInt64(); err != nil {
			return nil, err
		}
		if req.ProducerEpoch, err = r.ReadInt16(); err != nil {
			return nil, err
		}
	}
	if flexible {
		if err := r.ReadTaggedFields(); err != nil {
			return nil, err
		}
	}
	return req, nil
}

func EncodeInitProducerIdResponse(w *codec.Writer, resp *InitProducerIdResponse, version int16) {
	flexible := version >= 2
	w.WriteInt32(resp.ThrottleTimeMs)
	w.WriteInt16(resp.ErrorCode)
	w.WriteInt64(resp.ProducerID)
	w.WriteInt16(resp.ProducerEpoch)
	if flexible {
		w.WriteEmptyTaggedFields()
	}
}

File: crates/kaas-codec/src/api/init_producer_id.rs
//! InitProducerId — API key 22. v0–v4, flexible from v2 (KIP-482).
//!
//!
//! The Java producer issues this on startup when
//! `enable.idempotence=true` (the default since Kafka 3.0) to obtain a
//! `(producer_id, epoch)` pair used to tag every Produce batch for
//! the producer's lifetime.
//!
//! v3+ adds `(producer_id, producer_epoch)` request fields per KIP-360
//! so a producer can ask for renewal of an existing PID after a fatal
//! error. Kaas accepts but currently ignores these — every
//! `InitProducerId` returns a fresh PID with `epoch = 0`. The Phase 6
//! transaction work wires the gh #22 rejoin contract that respects the
//! request fields.

use bytes::BytesMut;

use crate::api::common::read_nullable_str;
use crate::api::registry::ApiSpec;
use crate::errors::CodecError;
use crate::headers::HeaderVersion;
use crate::primitives::{read_i16, read_i32, read_i64, write_i16, write_i32, write_i64};
use crate::tagged;
use crate::Bytes;

pub const VERSIONS: (i16, i16) = (0, 4);
pub const MIN_FLEXIBLE: i16 = 2;

const fn header_for(version: i16) -> HeaderVersion {
    if version >= MIN_FLEXIBLE {
        HeaderVersion::V2
    } else {
        HeaderVersion::V1
    }
}

fn request_hdr(version: i16) -> HeaderVersion {
    header_for(version)
}

fn response_hdr(version: i16) -> HeaderVersion {
    header_for(version)
}

pub const SPEC: ApiSpec = ApiSpec {
    key: 22,
    min_version: VERSIONS.0,
    max_version: VERSIONS.1,
    min_flexible: Some(MIN_FLEXIBLE),
    request_hdr,
    response_hdr,
};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Request {
    /// `None` ↔ wire null = non-transactional producer.
    /// `Some(id)` = transactional producer (Phase 6 honours the id).
    pub transactional_id: Option<String>,
    pub transaction_timeout_ms: i32,
    /// v3+. `-1` when not present (legacy versions).
    pub producer_id: i64,
    /// v3+. `-1` when not present.
    pub producer_epoch: i16,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Response {
    pub throttle_time_ms: i32,
    pub error_code: i16,
    pub producer_id: i64,
    pub producer_epoch: i16,
}

pub fn decode_request(buf: &mut Bytes, version: i16) -> Result<Request, CodecError> {
    let flexible = version >= MIN_FLEXIBLE;

    let transactional_id = read_nullable_str(buf, flexible)?;
    let transaction_timeout_ms = read_i32(buf)?;
    let (producer_id, producer_epoch) = if version >= 3 {
        (read_i64(buf)?, read_i16(buf)?)
    } else {
        (-1, -1)
    };
    if flexible {
        tagged::read(buf)?;
    }
    Ok(Request {
        transactional_id,
        transaction_timeout_ms,
        producer_id,
        producer_epoch,
    })
}

pub fn encode_response(
    buf: &mut BytesMut,
    resp: &Response,
    version: i16,
) -> Result<(), CodecError> {
    let flexible = version >= MIN_FLEXIBLE;
    write_i32(buf, resp.throttle_time_ms);
    write_i16(buf, resp.error_code);
    write_i64(buf, resp.producer_id);
    write_i16(buf, resp.producer_epoch);
    if flexible {
        tagged::write_empty(buf);
    }
    Ok(())
}

pub fn decode_response(buf: &mut Bytes, version: i16) -> Result<Response, CodecError> {
    let flexible = version >= MIN_FLEXIBLE;
    let throttle_time_ms = read_i32(buf)?;
    let error_code = read_i16(buf)?;
    let producer_id = read_i64(buf)?;
    let producer_epoch = read_i16(buf)?;
    if flexible {
        tagged::read(buf)?;
    }
    Ok(Response {
        throttle_time_ms,
        error_code,
        producer_id,
        producer_epoch,
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::api::common::write_nullable_str;

    fn encode_request(req: &Request, version: i16) -> BytesMut {
        let flexible = version >= MIN_FLEXIBLE;
        let mut w = BytesMut::new();
        write_nullable_str(&mut w, req.transactional_id.as_deref(), flexible).unwrap();
        write_i32(&mut w, req.transaction_timeout_ms);
        if version >= 3 {
            write_i64(&mut w, req.producer_id);
            write_i16(&mut w, req.producer_epoch);
        }
        if flexible {
            tagged::write_empty(&mut w);
        }
        w
    }

    fn sample_request(version: i16) -> Request {
        Request {
            transactional_id: Some("tx-1".to_owned()),
            transaction_timeout_ms: 60_000,
            producer_id: if version >= 3 { 12345 } else { -1 },
            producer_epoch: if version >= 3 { 7 } else { -1 },
        }
    }

    fn roundtrip_request(version: i16) {
        let req = sample_request(version);
        let w = encode_request(&req, version);
        let mut r = w.freeze();
        let got = decode_request(&mut r, version).unwrap();
        assert_eq!(got, req, "request v{version}");
        assert!(r.is_empty());
    }

    fn roundtrip_response(version: i16) {
        let resp = Response {
            throttle_time_ms: 0,
            error_code: 0,
            producer_id: 999,
            producer_epoch: 0,
        };
        let mut w = BytesMut::new();
        encode_response(&mut w, &resp, version).unwrap();
        let mut r = w.freeze();
        let got = decode_response(&mut r, version).unwrap();
        assert_eq!(got, resp, "response v{version}");
        assert!(r.is_empty());
    }

    #[test]
    fn request_v0_roundtrip() {
        roundtrip_request(0);
    }
    #[test]
    fn request_v2_flexible_roundtrip() {
        roundtrip_request(2);
    }
    #[test]
    fn request_v3_pid_fields_present() {
        roundtrip_request(3);
    }
    #[test]
    fn request_v4_roundtrip() {
        roundtrip_request(4);
    }

    #[test]
    fn response_v0_roundtrip() {
        roundtrip_response(0);
    }
    #[test]
    fn response_v2_flexible_roundtrip() {
        roundtrip_response(2);
    }
    #[test]
    fn response_v4_roundtrip() {
        roundtrip_response(4);
    }

    #[test]
    fn null_transactional_id_is_non_transactional() {
        let req = Request {
            transactional_id: None,
            transaction_timeout_ms: 0,
            producer_id: -1,
            producer_epoch: -1,
        };
        let w = encode_request(&req, 4);
        let mut r = w.freeze();
        let got = decode_request(&mut r, 4).unwrap();
        assert_eq!(got.transactional_id, None);
    }
}
